File: ios/Sources/Account/Models/SubscriptionModels.swift
import Foundation

// MARK: - Subscription Plan

/// User subscription plans with associated limits and features.
enum Plan: String, Codable, CaseIterable, Sendable {
    case free
    case pro
    case unlimited

    // MARK: - Display Properties

    var displayName: String {
        switch self {
        case .free: return "Free"
        case .pro: return "Pro"
        case .unlimited: return "Unlimited"
        }
    }

    var shortDescription: String {
        switch self {
        case .free: return "Basic access to cloud voices"
        case .pro: return "Full access with higher limits"
        case .unlimited: return "Unlimited cloud voice usage"
        }
    }

    var badgeColor: String {
        switch self {
        case .free: return "#6B7280"      // Gray
        case .pro: return "#8B5CF6"       // Purple
        case .unlimited: return "#F59E0B" // Amber
        }
    }

    var iconName: String {
        switch self {
        case .free: return "person.circle"
        case .pro: return "star.circle.fill"
        case .unlimited: return "crown.fill"
        }
    }

    // MARK: - Limits

    /// Monthly character limit for cloud TTS
    var monthlyCharacterLimit: Int {
        switch self {
        case .free: return 25_000          // 25K chars (~30 min audio total per month)
        case .pro: return 1_000_000        // 1M chars (~12 hours audio)
        case .unlimited: return Int.max    // Unlimited
        }
    }

    /// Daily character limit for cloud TTS
    var dailyCharacterLimit: Int {
        switch self {
        case .free: return 4_000           // 4K chars (~5 min per day)
        case .pro: return 100_000          // 100K chars (~70 min)
        case .unlimited: return Int.max    // Unlimited
        }
    }

    /// Maximum article length for import
    var maxArticleLength: Int {
        switch self {
        case .free: return 10_000          // ~10K chars
        case .pro: return 50_000           // ~50K chars
        case .unlimited: return Int.max    // Unlimited
        }
    }

    /// Number of saved articles allowed
    var maxSavedArticles: Int {
        switch self {
        case .free: return 10
        case .pro: return 100
        case .unlimited: return Int.max
        }
    }

    /// Whether premium voices are accessible
    var hasPremiumVoiceAccess: Bool {
        switch self {
        case .free: return false
        case .pro, .unlimited: return true
        }
    }

    /// Whether character voices (Santa, etc.) are accessible
    var hasCharacterVoiceAccess: Bool {
        switch self {
        case .free: return false
        case .pro, .unlimited: return true
        }
    }

    /// Whether API access is available
    var hasAPIAccess: Bool {
        switch self {
        case .free: return false
        case .pro, .unlimited: return true
        }
    }

    /// Priority level for synthesis queue (higher = more priority)
    var synthesisPriority: Int {
        switch self {
        case .free: return 1
        case .pro: return 5
        case .unlimited: return 10
        }
    }

    // MARK: - Formatting

    var formattedMonthlyLimit: String {
        if monthlyCharacterLimit == Int.max {
            return "Unlimited"
        }
        return formatCharacterCount(monthlyCharacterLimit)
    }

    var formattedDailyLimit: String {
        if dailyCharacterLimit == Int.max {
            return "Unlimited"
        }
        return formatCharacterCount(dailyCharacterLimit)
    }

    // MARK: - Comparison

    /// Check if this plan has at least the features of another plan
    func hasAccess(to requiredPlan: Plan) -> Bool {
        let planOrder: [Plan] = [.free, .pro, .unlimited]
        guard let selfIndex = planOrder.firstIndex(of: self),
              let requiredIndex = planOrder.firstIndex(of: requiredPlan) else {
            return false
        }
        return selfIndex >= requiredIndex
    }

    /// The minimum plan required for a feature
    static func minimumPlanFor(feature: PlanFeature) -> Plan {
        switch feature {
        case .basicVoices, .onDeviceTTS:
            return .free
        case .premiumVoices, .characterVoices, .extendedArticles, .apiAccess:
            return .pro
        case .unlimitedUsage:
            return .unlimited
        }
    }
}

// MARK: - Plan Features

/// Features that require specific plans
enum PlanFeature: String, CaseIterable, Sendable {
    case basicVoices
    case premiumVoices
    case characterVoices
    case onDeviceTTS
    case extendedArticles
    case apiAccess
    case unlimitedUsage

    var displayName: String {
        switch self {
        case .basicVoices: return "Basic Cloud Voices"
        case .premiumVoices: return "Premium Voices"
        case .characterVoices: return "Character Voices"
        case .onDeviceTTS: return "On-Device TTS"
        case .extendedArticles: return "Extended Articles"
        case .apiAccess: return "API Access"
        case .unlimitedUsage: return "Unlimited Usage"
        }
    }

    var description: String {
        switch self {
        case .basicVoices: return "Access to standard quality cloud voices"
        case .premiumVoices: return "Access to premium ElevenLabs and OpenAI voices"
        case .characterVoices: return "Fun character voices like Santa and celebrities"
        case .onDeviceTTS: return "Offline text-to-speech using Apple voices"
        case .extendedArticles: return "Import and listen to longer articles"
        case .apiAccess: return "Programmatic access to TTS via API"
        case .unlimitedUsage: return "No monthly character limits"
        }
    }
}

// MARK: - Subscription Status

/// Current subscription status
enum SubscriptionStatus: String, Codable, Sendable {
    case active
    case trialing
    case pastDue
    case canceled
    case expired
    case none

    var isActive: Bool {
        switch self {
        case .active, .trialing:
            return true
        case .pastDue, .canceled, .expired, .none:
            return false
        }
    }

    var displayName: String {
        switch self {
        case .active: return "Active"
        case .trialing: return "Trial"
        case .pastDue: return "Past Due"
        case .canceled: return "Canceled"
        case .expired: return "Expired"
        case .none: return "None"
        }
    }

    var statusColor: String {
        switch self {
        case .active: return "#10B981"     // Green
        case .trialing: return "#3B82F6"   // Blue
        case .pastDue: return "#F59E0B"    // Amber
        case .canceled, .expired: return "#EF4444" // Red
        case .none: return "#6B7280"       // Gray
        }
    }
}

// MARK: - User Profile

/// User profile with subscription information
struct UserProfile: Codable, Sendable {
    let id: String
    let email: String?
    let displayName: String?
    let avatarURL: URL?
    let plan: Plan
    let subscriptionStatus: SubscriptionStatus
    let subscriptionExpiresAt: Date?
    let createdAt: Date
    let updatedAt: Date

    // MARK: - Computed Properties

    var effectivePlan: Plan {
        // If subscription is not active, fall back to free
        guard subscriptionStatus.isActive else {
            return .free
        }
        return plan
    }

    var isSubscribed: Bool {
        plan != .free && subscriptionStatus.isActive
    }

    var daysUntilExpiration: Int? {
        guard let expiresAt = subscriptionExpiresAt else { return nil }
        let days = Calendar.current.dateComponents([.day], from: Date(), to: expiresAt).day
        return days
    }

    var isExpiringSoon: Bool {
        guard let days = daysUntilExpiration else { return false }
        return days <= 7 && days > 0
    }

    // MARK: - Initialization

    init(
        id: String,
        email: String? = nil,
        displayName: String? = nil,
        avatarURL: URL? = nil,
        plan: Plan = .free,
        subscriptionStatus: SubscriptionStatus = .none,
        subscriptionExpiresAt: Date? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.email = email
        self.displayName = displayName
        self.avatarURL = avatarURL
        self.plan = plan
        self.subscriptionStatus = subscriptionStatus
        self.subscriptionExpiresAt = subscriptionExpiresAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    // MARK: - Feature Access

    func canAccess(feature: PlanFeature) -> Bool {
        let requiredPlan = Plan.minimumPlanFor(feature: feature)
        return effectivePlan.hasAccess(to: requiredPlan)
    }

    func canAccessVoice(tier: VoiceTier) -> Bool {
        switch tier {
        case .free:
            return true
        case .premium:
            return effectivePlan.hasPremiumVoiceAccess
        case .enterprise:
            return effectivePlan.hasCharacterVoiceAccess
        }
    }
}

// MARK: - Subscription Info Response

/// Response from backend /usage/full endpoint
struct FullUsageResponse: Codable, Sendable {
    let subscription: SubscriptionInfo
    let usage: UsageDetails
    let estimatedMinutesRemaining: Int

    enum CodingKeys: String, CodingKey {
        case subscription
        case usage
        case estimatedMinutesRemaining = "estimated_minutes_remaining"
    }

    struct SubscriptionInfo: Codable, Sendable {
        let tier: String
        let status: String
        let currentPeriodEnd: String?

        enum CodingKeys: String, CodingKey {
            case tier
            case status
            case currentPeriodEnd = "current_period_end"
        }
    }

    struct UsageDetails: Codable, Sendable {
        let daily: QuotaInfo
        let monthly: QuotaInfo
    }

    struct QuotaInfo: Codable, Sendable {
        let used: Int
        let limit: Int
        let remaining: Int
        let percentage: Double
        let resetsAt: String

        enum CodingKeys: String, CodingKey {
            case used
            case limit
            case remaining
            case percentage
            case resetsAt = "resets_at"
        }
    }

    // MARK: - Convenience Accessors

    var planEnum: Plan {
        Plan(rawValue: subscription.tier) ?? .free
    }

    var statusEnum: SubscriptionStatus {
        SubscriptionStatus(rawValue: subscription.status) ?? .none
    }

    var expiresAtDate: Date? {
        guard let dateString = subscription.currentPeriodEnd else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: dateString) {
            return date
        }
        // Try without fractional seconds
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: dateString)
    }

    var monthlyCharUsed: Int {
        usage.monthly.used
    }

    var monthlyCharLimit: Int {
        usage.monthly.limit
    }

    var dailyCharUsed: Int {
        usage.daily.used
    }

    var dailyCharLimit: Int {
        usage.daily.limit
    }
}

/// Response from subscription sync endpoint
struct SubscriptionSyncResponse: Codable, Sendable {
    let success: Bool
    let subscription: SyncedSubscription
    let limits: SyncedLimits

    struct SyncedSubscription: Codable, Sendable {
        let tier: String
        let status: String
        let expiresAt: String?

        enum CodingKeys: String, CodingKey {
            case tier
            case status
            case expiresAt = "expires_at"
        }
    }

    struct SyncedLimits: Codable, Sendable {
        let daily: Int
        let monthly: Int
    }

    var planEnum: Plan {
        Plan(rawValue: subscription.tier) ?? .free
    }

    var statusEnum: SubscriptionStatus {
        SubscriptionStatus(rawValue: subscription.status) ?? .none
    }
}

/// Legacy response format (kept for compatibility)
struct SubscriptionInfoResponse: Codable, Sendable {
    let userId: String
    let plan: String
    let status: String
    let expiresAt: String?
    let monthlyCharUsed: Int
    let monthlyCharLimit: Int

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case plan
        case status
        case expiresAt = "expires_at"
        case monthlyCharUsed = "monthly_char_used"
        case monthlyCharLimit = "monthly_char_limit"
    }

    var planEnum: Plan {
        Plan(rawValue: plan) ?? .free
    }

    var statusEnum: SubscriptionStatus {
        SubscriptionStatus(rawValue: status) ?? .none
    }

    var expiresAtDate: Date? {
        guard let expiresAt = expiresAt else { return nil }
        let formatter = ISO8601DateFormatter()
        return formatter.date(from: expiresAt)
    }

    /// Create from FullUsageResponse
    init(from fullResponse: FullUsageResponse, userId: String) {
        self.userId = userId
        self.plan = fullResponse.subscription.tier
        self.status = fullResponse.subscription.status
        self.expiresAt = fullResponse.subscription.currentPeriodEnd
        self.monthlyCharUsed = fullResponse.monthlyCharUsed
        self.monthlyCharLimit = fullResponse.monthlyCharLimit
    }

    init(
        userId: String,
        plan: String,
        status: String,
        expiresAt: String?,
        monthlyCharUsed: Int,
        monthlyCharLimit: Int
    ) {
        self.userId = userId
        self.plan = plan
        self.status = status
        self.expiresAt = expiresAt
        self.monthlyCharUsed = monthlyCharUsed
        self.monthlyCharLimit = monthlyCharLimit
    }
}

File: android/app/src/main/java/com/listenai/data/models/SubscriptionModels.kt
package com.listenai.data.models

import java.time.Instant
import java.time.temporal.ChronoUnit

/**
 * User subscription plans with associated limits and features.
 */
enum class Plan(val displayName: String) {
    FREE("Free"),
    PRO("Pro"),
    UNLIMITED("Unlimited");

    val shortDescription: String
        get() = when (this) {
            FREE -> "Basic access to cloud voices"
            PRO -> "Full access with higher limits"
            UNLIMITED -> "Unlimited cloud voice usage"
        }

    val badgeColor: Long
        get() = when (this) {
            FREE -> 0xFF6B7280      // Gray
            PRO -> 0xFF8B5CF6       // Purple
            UNLIMITED -> 0xFFF59E0B // Amber
        }

    val iconName: String
        get() = when (this) {
            FREE -> "person"
            PRO -> "star"
            UNLIMITED -> "crown"
        }

    // Limits
    val monthlyCharacterLimit: Int
        get() = when (this) {
            FREE -> 25_000          // 25K chars (~30 min audio total per month)
            PRO -> 1_000_000        // 1M chars (~12 hours audio)
            UNLIMITED -> Int.MAX_VALUE
        }

    val dailyCharacterLimit: Int
        get() = when (this) {
            FREE -> 4_000           // 4K chars (~5 min per day)
            PRO -> 100_000          // 100K chars (~70 min)
            UNLIMITED -> Int.MAX_VALUE
        }

    val maxArticleLength: Int
        get() = when (this) {
            FREE -> 10_000
            PRO -> 50_000
            UNLIMITED -> Int.MAX_VALUE
        }

    val maxSavedArticles: Int
        get() = when (this) {
            FREE -> 10
            PRO -> 100
            UNLIMITED -> Int.MAX_VALUE
        }

    val hasPremiumVoiceAccess: Boolean
        get() = this != FREE

    val hasCharacterVoiceAccess: Boolean
        get() = this != FREE

    val hasAPIAccess: Boolean
        get() = this != FREE

    val synthesisPriority: Int
        get() = when (this) {
            FREE -> 1
            PRO -> 5
            UNLIMITED -> 10
        }

    val formattedMonthlyLimit: String
        get() = if (monthlyCharacterLimit == Int.MAX_VALUE) {
            "Unlimited"
        } else {
            formatCharacterCount(monthlyCharacterLimit)
        }

    val formattedDailyLimit: String
        get() = if (dailyCharacterLimit == Int.MAX_VALUE) {
            "Unlimited"
        } else {
            formatCharacterCount(dailyCharacterLimit)
        }

    fun hasAccess(requiredPlan: Plan): Boolean {
        val planOrder = listOf(FREE, PRO, UNLIMITED)
        return planOrder.indexOf(this) >= planOrder.indexOf(requiredPlan)
    }

    companion object {
        fun fromString(value: String): Plan {
            return entries.find { it.name.equals(value, ignoreCase = true) } ?: FREE
        }

        fun minimumPlanFor(feature: PlanFeature): Plan {
            return when (feature) {
                PlanFeature.BASIC_VOICES, PlanFeature.ON_DEVICE_TTS -> FREE
                PlanFeature.PREMIUM_VOICES, PlanFeature.CHARACTER_VOICES,
                PlanFeature.EXTENDED_ARTICLES, PlanFeature.API_ACCESS -> PRO
                PlanFeature.UNLIMITED_USAGE -> UNLIMITED
            }
        }
    }
}

/**
 * Features that require specific plans
 */
enum class PlanFeature(val displayName: String, val description: String) {
    BASIC_VOICES("Cloud Voices", "Access to high-quality cloud voices"),
    PREMIUM_VOICES("Premium Voices", "Access to all premium AI voices"),
    CHARACTER_VOICES("Character Voices", "Fun character voices like Santa and celebrities"),
    ON_DEVICE_TTS("On-Device TTS", "Offline text-to-speech using system voices"),
    EXTENDED_ARTICLES("Extended Articles", "Import and listen to longer articles"),
    API_ACCESS("API Access", "Programmatic access to TTS via API"),
    UNLIMITED_USAGE("Unlimited Usage", "No monthly character limits")
}

/**
 * Current subscription status
 */
enum class SubscriptionStatus(val displayName: String) {
    ACTIVE("Active"),
    TRIALING("Trial"),
    PAST_DUE("Past Due"),
    CANCELED("Canceled"),
    EXPIRED("Expired"),
    NONE("None");

    val isActive: Boolean
        get() = this == ACTIVE || this == TRIALING

    val statusColor: Long
        get() = when (this) {
            ACTIVE -> 0xFF10B981     // Green
            TRIALING -> 0xFF3B82F6   // Blue
            PAST_DUE -> 0xFFF59E0B   // Amber
            CANCELED, EXPIRED -> 0xFFEF4444 // Red
            NONE -> 0xFF6B7280       // Gray
        }

    companion object {
        fun fromString(value: String): SubscriptionStatus {
            return entries.find { it.name.equals(value, ignoreCase = true) } ?: NONE
        }
    }
}

/**
 * User profile with subscription information
 */
data class UserProfile(
    val id: String,
    val email: String? = null,
    val displayName: String? = null,
    val avatarUrl: String? = null,
    val plan: Plan = Plan.FREE,
    val subscriptionStatus: SubscriptionStatus = SubscriptionStatus.NONE,
    val subscriptionExpiresAt: Instant? = null,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now()
) {
    val effectivePlan: Plan
        get() = if (subscriptionStatus.isActive) plan else Plan.FREE

    val isSubscribed: Boolean
        get() = plan != Plan.FREE && subscriptionStatus.isActive

    val daysUntilExpiration: Int?
        get() = subscriptionExpiresAt?.let {
            ChronoUnit.DAYS.between(Instant.now(), it).toInt()
        }

    val isExpiringSoon: Boolean
        get() = daysUntilExpiration?.let { it in 1..7 } ?: false

    fun canAccess(feature: PlanFeature): Boolean {
        val requiredPlan = Plan.minimumPlanFor(feature)
        return effectivePlan.hasAccess(requiredPlan)
    }

    fun canAccessVoice(tier: VoiceTier): Boolean {
        return when (tier) {
            VoiceTier.FREE -> true
            VoiceTier.PREMIUM -> effectivePlan.hasPremiumVoiceAccess
            VoiceTier.ENTERPRISE -> effectivePlan.hasCharacterVoiceAccess
        }
    }

    companion object {
        val ANONYMOUS = UserProfile(
            id = "anonymous",
            plan = Plan.FREE,
            subscriptionStatus = SubscriptionStatus.NONE
        )
    }
}

/**
 * Response from backend /usage/full endpoint
 */
data class FullUsageResponse(
    val subscription: SubscriptionInfo,
    val usage: UsageDetails,
    val estimatedMinutesRemaining: Int
) {
    data class SubscriptionInfo(
        val tier: String,
        val status: String,
        val currentPeriodEnd: String?
    )

    data class UsageDetails(
        val daily: QuotaInfo,
        val monthly: QuotaInfo
    )

    data class QuotaInfo(
        val used: Int,
        val limit: Int,
        val remaining: Int,
        val percentage: Double,
        val resetsAt: String
    )

    val planEnum: Plan
        get() = Plan.fromString(subscription.tier)

    val statusEnum: SubscriptionStatus
        get() = SubscriptionStatus.fromString(subscription.status)

    val expiresAtInstant: Instant?
        get() = subscription.currentPeriodEnd?.let {
            try {
                Instant.parse(it)
            } catch (e: Exception) {
                null
            }
        }

    val monthlyCharUsed: Int
        get() = usage.monthly.used

    val monthlyCharLimit: Int
        get() = usage.monthly.limit

    val dailyCharUsed: Int
        get() = usage.daily.used

    val dailyCharLimit: Int
        get() = usage.daily.limit
}

/**
 * Response from subscription sync endpoint
 */
data class SubscriptionSyncResponse(
    val success: Boolean,
    val subscription: SyncedSubscription,
    val limits: SyncedLimits
) {
    data class SyncedSubscription(
        val tier: String,
        val status: String,
        val expiresAt: String?
    )

    data class SyncedLimits(
        val daily: Int,
        val monthly: Int
    )

    val planEnum: Plan
        get() = Plan.fromString(subscription.tier)

    val statusEnum: SubscriptionStatus
        get() = SubscriptionStatus.fromString(subscription.status)
}

/**
 * Helper function to format character counts
 */
private fun formatCharacterCount(count: Int): String {
    return when {
        count >= 1_000_000 -> "${count / 1_000_000}M"
        count >= 1_000 -> "${count / 1_000}K"
        else -> count.toString()
    }
}
